sdk/tests/core/test_az_mqtt5_connection.c

```diff
@@ -434,28 +668,917 @@ static void test_az_mqtt5_connection_open_idle_success(void** state)
+                                    .tls_authentication_error = false }),
+      AZ_OK);
+
+  assert_int_equal(ref_conn_rsp_ok, 2);
+  assert_int_equal(ref_conn_rsp_err, 0);
+  assert_int_equal(ref_conn_error, 0);
+  assert_int_equal(ref_disconn_req, 1);
+}
+
+// Testing that when disconnecting, if a disconnect response is not received before the timeout, the
+// connection goes into a faulted state.
+static void test_az_mqtt5_connection_disconnecting_timeout_failure(void** state)
+{
+  (void)state;
+  _reset_refs();
+  az_mqtt5_connection test_connection;
+  az_mqtt5 mock_mqtt;
+  az_mqtt5_options mock_mqtt_options = { 0 };
+  _az_event_client mock_client_1;
+  _az_event_client mock_client_2;
+  _az_hfsm mock_client_hfsm_1;
+  _az_hfsm mock_client_hfsm_2;
+
+  _az_mqtt5_connection_setup(
+      &test_connection,
+      &test_connection_options,
+      &mock_mqtt,
+      &mock_mqtt_options,
+      &mock_client_1,
+      &mock_client_2,
+      &mock_client_hfsm_1,
+      &mock_client_hfsm_2);
+
+  _az_mqtt5_connection_idle_connecting(&test_connection);
+  _az_mqtt5_connection_connecting_connected(&mock_mqtt);
+  _az_mqtt5_connection_connected_disconnecting(&test_connection);
+
+  _az_event_pipeline_timer* timer = &test_connection._internal.connection_timer;
+  az_event timer_event = (az_event){ .type = AZ_HFSM_EVENT_TIMEOUT, .data = timer };
+
+  assert_int_equal(
+      _az_event_pipeline_post_outbound_event(timer->_internal.pipeline, timer_event), AZ_OK);
+
+  assert_int_equal(ref_timeout, 1);
+  assert_int_equal(ref_conn_rsp_ok, 2);
+  assert_int_equal(ref_conn_rsp_err, 0);
+  assert_int_equal(ref_conn_error, 1);
+  assert_int_equal(ref_disconn_req, 1);
+
+  _az_mqtt5_connection_faulted_check(&test_connection);
+}
+
+// Testing that when waiting to reconnect, an attempt to open the connection is ignored.
+static void test_az_mqtt5_connection_reconnecttimeout_ignore_success(void** state)
+{
+  (void)state;
+  _reset_refs();
+  az_mqtt5_connection test_connection;
+  az_mqtt5 mock_mqtt;
+  az_mqtt5_options mock_mqtt_options = { 0 };
+  _az_event_client mock_client_1;
+  _az_event_client mock_client_2;
+  _az_hfsm mock_client_hfsm_1;
+  _az_hfsm mock_client_hfsm_2;
+
+  _az_mqtt5_connection_setup(
+      &test_connection,
+      &test_connection_options,
+      &mock_mqtt,
+      &mock_mqtt_options,
+      &mock_client_1,
+      &mock_client_2,
+      &mock_client_hfsm_1,
+      &mock_client_hfsm_2);
+
+  _az_mqtt5_connection_idle_connecting(&test_connection);
+  _az_mqtt5_connection_connecting_reconnecttimeout(&mock_mqtt, AZ_MQTT5_CONNACK_UNSPECIFIED_ERROR);
+
+  az_mqtt5_connection_open(&test_connection);
+  assert_int_equal(ref_conn_rsp_ok, 0);
+  assert_int_equal(ref_conn_rsp_err, 1);
+  assert_int_equal(ref_conn_req, 2);
+}
+
+// Testing that when reconnecting, an ok connect response from the broker leads to a faulted state.
+static void test_az_mqtt5_connection_reconnecttimeout_connect_rsp_ok_failure(void** state)
```

Comment:
Shouldn't an ok connect response during a reconnect be what we're hoping for? (instead of going to a faulted state)

Reply:
We arrived at the error because of a connect response error so it would be odd for us to receive a connect response ok when we haven't sent out another connect request yet